### PR TITLE
Login via POST method and default scopes

### PR DESCRIPTION
**Pull request by ibimsdeandy**
Login via POST method
Use default scopes when no scopes are given

### CODE REVIEW

File: src/Control/Controller.php [hunk 1]
```diff
@@ -131,6 +138,10 @@ public function callback(HTTPRequest $request)
             $session->clear('oauth2');
             $this->httpError(400, 'Invalid session state.');
             return null;
+        } else if ($this->validateState($request) === 'isPost') {
+            return $this->renderWith('OAuthRedirect', [
+                'url' => sprintf("/%s?code=%s&state=%s", $request->getURL(), $request->postVar('code'), $request->postVar('state'))
```
Comment (kinglozzer) [inline src/Control/Controller.php:143]: Isn’t this redirecting back to the same URL, just a GET instead of a POST? Why does this render a template that then redirects, instead of performing a redirect directly?

Comment (ibimsdeandy) [reply src/Control/Controller.php]: Apple makes a POST instead of a GET request which leads to CORS problems. This intermediate step was introduced to make it a GET request.

$this->redirect() would not work because the request has to happen on the client side (via JavaScript).

File: src/Control/Controller.php [hunk 2]
```diff
@@ -262,6 +273,10 @@ public function validateState(HTTPRequest $request)
         $session = $request->getSession();
         $data = $session->get('oauth2');
 
+        if ($request->postVar('state') && $request->postVar('code')) {
+            return 'isPost';
```
[the remark this answers is not in the document]
Comment (ibimsdeandy) [reply src/Control/Controller.php]: We moved the if from the validateState function to the callback function. This step is required because Apple uses POST instead of GET requests

File: src/Control/Controller.php [hunk 3]
```diff
@@ -94,15 +94,22 @@ public function authenticate(HTTPRequest $request)
         $context = $request->getVar('context');
         $scope = $request->getVar('scope');
 
+
         // Missing or invalid data means we can't proceed
-        if (!$providerName || !is_array($scope)) {
+        if (!$providerName || !empty($scope) && !is_array($scope)) {
```
[the remark this answers is not in the document]
Comment (ibimsdeandy) [reply src/Control/Controller.php]: Apple does not supply a scope. However, the check has been done in the meantime. It was a leftover from the trial and error.

File: src/Control/Controller.php [hunk 4]
```diff
@@ -94,15 +94,22 @@ public function authenticate(HTTPRequest $request)
...
             return null;
         }
 
         /** @var ProviderFactory $providerFactory */
         $providerFactory = Injector::inst()->get(ProviderFactory::class);
         $provider = $providerFactory->getProvider($providerName);
+
+        // Ensure we always have scope to work with
+        if (empty($scope)) {
+            $scope = $provider->defaultScopes;
```
[the remark this answers is not in the document]
Comment (ibimsdeandy) [reply src/Control/Controller.php]: getDefaultScopes is a protected function in the Apple Provider. Therefore we got the value directly from the variable. I have adjusted the line. Additionally I will open a pull request at the plugin patrickbussmann/oauth2-apple to set the function to public.

**Comment by johannesx75** [comment]
> Do you have an example of a provider that requires these changes? It’s hard to follow why these changes are being made without that

Sign in with Apple needs them. Unfortunately they decided to use post instead of get to bounce back to the origin server. We used [this provider](https://github.com/patrickbussmann/oauth2-apple/) for our implementation.

The good news is, it passed Apples App Store Review. You can see it in action [here](https://www.werstreamt.es/Security/login).

**Comment by kinglozzer** [comment]
Thanks @ibimsdeandy @johannesx75, it’s starting to make a bit more sense now.

If the only issue is that Apple use a POST request instead of a GET request, could we undo the changes that add the redirect (the `OAuthRedirect` template etc) and instead change these lines to use `requestVar()` instead of `getVar()`?

https://github.com/bigfork/silverstripe-oauth/blob/88704d6e26758c675f88ed85c310751b3a59fd39/src/Control/Controller.php#L164C37-L164C43

https://github.com/bigfork/silverstripe-oauth/blob/88704d6e26758c675f88ed85c310751b3a59fd39/src/Control/Controller.php#L281C28-L281C34

It seems to me that this would allow POST requests to work without adding a redirect. What do you think?

**Comment by johannesx75** [comment]
Hi @kinglozzer,

just changing the code to accept POST had been our first idea as well. Unfortunately that does not work, in some cases.

The reason is, that you don't have access to the session in those requests then. This is because of the SameSite setting on the Session Cookie. If it is not set, the browser will default to Lax. In that setting Chrome (and some others) will not sent cookies on cross-site requests with POST. To get around this, we take the data and create a same domain GET request, that is triggered by the clients browser.

**Comment by kinglozzer** [comment]
Ah okay, so the redirect is essentially a workaround to avoid setting `SameSite=None; Secure`? I.e. the use-case described [here](https://web.dev/samesite-cookie-recipes/#unsafe-requests-across-sites)

**Comment by johannesx75** [comment]
> Ah okay, so the redirect is essentially a workaround to avoid setting `SameSite=None; Secure`? I.e. the use-case described [here](https://web.dev/samesite-cookie-recipes/#unsafe-requests-across-sites)

Yes!

**Comment by kinglozzer** [comment]
Sorry it’s taken me a while to look at this. As I don’t have an account/app set up with Apple, could you test [this branch](https://github.com/bigfork/silverstripe-oauth/tree/samesite-workaround)? You’ll need to set a config flag to enable this: https://github.com/bigfork/silverstripe-oauth/tree/samesite-workaround#samesite--post-cookies

Thanks!

**Comment by johannesx75** [comment]
Hi @kinglozzer, the branch works perfect. As long as you use our fork of the Apple Oauth2 Provider: https://github.com/werstreamtes/oauth2-apple

The reason is, that in patrickbussmann/oauth2-apple the getDefaultScopes is protected. We have issued a Pull Request, but so far he hasn't responded. 

So maybe you want to add a little note regarding setting up the composer file to  the readme.
We use:
`"patrickbussmann/oauth2-apple" : "dev-main"`
And then add:

```
    "repositories": [
        {
            "name": "patrickbussmann/oauth2-apple",
            "type": "vcs",
            "url": "https://github.com/werstreamtes/oauth2-apple"
        }
]
```

You can see it in action at: https://www.werstreamt.es/Security/login?BackURL=%2F
